[Buildpacks] Support env variables

dgageot opened the pull request:
It's now possible to pass custom env variables to the buildpacks.
This is required to pass some configuration to the buildpacks.

Signed-off-by: David Gageot <david@gageot.net>

dgageot commented:
Sure! I'm planning on adding some doc for the buildpacks support anyways

briandealwis commented on docs/content/en/docs/pipeline-stages/builders/buildpacks.md:
```diff
@@ -25,9 +25,17 @@ The following options can optionally be configured:
+
+`env` makes it possible to pass arbitrary environment variables to the buildpacks.
+This is in general used to adjust the detection and the build phases. For eg., most
+builders let you change runtime versions with environment variables.
```
I think it’s buildpacks, not builders, that use the variables to restrict runtimes versions.

Wdyt: _Many buildpacks use environment variables to select specific versions of language runtimes._

dgageot replied:
Thanks. I'll take your text

briandealwis commented on docs/content/en/docs/pipeline-stages/builders/custom.md:
```diff
@@ -9,17 +9,19 @@ Custom build scripts allow skaffold users the flexibility to build artifacts wit
 | $PUSH_IMAGE      | Set to true if the image in `$IMAGE` is expected to exist in a remote registry. Set to false if the image is expected to exist locally.      |   The custom build script will push the image `$IMAGE` if `$PUSH_IMAGE=true` | 
-| $BUILD_CONTEXT  | An absolute path to the directory this artifact is meant to be built from. Specified by artifact `context` in the skaffold.yaml.      | None. | 
-| Local environment variables | The current state of the local environment (e.g. `$HOST`, `$PATH)`. Determined by the golang [os.Environ](https://golang.org/pkg/os#Environ) function.| None. |
+| $BUILD_CONTEXT  | An absolute path to the directory this artifact is meant to be built from. Specified by artifact `context` in the skaffold.yaml.      | |
```
This change in order seems to indicate the precedence?  Might be worth explicitly calling that out?

dgageot replied:
I've re-focused this PR on the buildpacks builder only

briandealwis commented on pkg/skaffold/build/custom/custom.go:
```diff
@@ -89,20 +89,27 @@ func (b *ArtifactBuilder) retrieveCmd(out io.Writer, a *latest.Artifact, tag str
+	var envs []string
+	envs = append(envs, util.OSEnviron()...)
+	envs = append(envs, b.additionalEnv...)
+	envs = append(envs, evaluatedEnv...)
```
Is there an implied order here?  I'd think that later values will override earlier values in the array, such that this means that values in the `skaffold.yaml` override those on the command-line?  Which seems unusual.

(I wonder if this function might not be better converting to a map and then back to a string array?)

briandealwis commented on pkg/skaffold/build/custom/custom.go:
```diff
@@ -89,20 +89,27 @@ func (b *ArtifactBuilder) retrieveCmd(out io.Writer, a *latest.Artifact, tag str
 
-func (b *ArtifactBuilder) retrieveEnv(a *latest.Artifact, tag string) ([]string, error) {
-	buildContext, err := buildContext(a.Workspace)
+func (b *ArtifactBuilder) retrieveEnv(workspace string, a *latest.CustomArtifact, tag string) ([]string, error) {
```
This name is at odds with `hash.go`'s `retrieveEnv` that literally just retrieves the appropriate env section.  This `retrieveEnv` doesn't seem quite right since there is generation happening here too.  `builderEnv()`?

dgageot replied:
I've re-focused this PR on the buildpacks builder only

briandealwis commented on pkg/skaffold/build/custom/custom_test.go:
```diff
@@ -30,36 +30,61 @@ func TestRetrieveEnv(t *testing.T) {
+				Env: []string{"INVALID"},
+			},
+			tag:       "gcr.io/image/push:tag",
+			shouldErr: true,
```
I'd like to see a test with overrides, if that's possible.

dgageot replied:
I've re-focused this PR on the buildpacks builder only

briandealwis commented on pkg/skaffold/build/misc/env_test.go:
```diff
@@ -0,0 +1,61 @@
+	}{
+		{
+			description: "env variables",
+			env:         []string{"key1=value1", "key2=", "key3={{.FOO}}"},
```
Should variables be able to refer to other variables in the same level?  For example, `key4={{.key3}}`?

dgageot replied:
Any example why we'd want to do that?

briandealwis replied:
I think go templates can be nested? https://golang.org/pkg/text/template/#hdr-Nested_template_definitions We should document any differences from normal usage of go templates, especially for the go-n00bs (like me!) who don't really know the limitations and how these uses would differ from other uses of go templates.

dgageot replied:
I'd rather keep it simple and not go into too much details about go templates.
I added an example to the `buildpacks` sample so that users understand the basic usage.

dgageot commented on docs/content/en/docs/pipeline-stages/builders/buildpacks.md:
```diff
@@ -61,6 +69,10 @@ buildpack:
 
+### File Sync
+
+Syncable files must be included in both the `paths` section of `dependencies`, so that the skaffold file watcher knows to watch them, and the `sync` section, so that skaffold knows how to sync them.
```
It really depends on the buildpack. Not a lot we can do to help I guess.

briandealwis replied:
Buildpacks might be a good candidate for the "smart sync".  Maybe we could have our buildpack record sync info in the buildpack metadata.

dgageot commented:
@balopat @briandealwis Thanks for your feedback. I've added the doc and removed the code related to the custom builder. I prefer to keep this PR focused on the buildpacks builder only.

briandealwis commented on docs/content/en/docs/pipeline-stages/builders/buildpacks.md:
```diff
@@ -25,9 +25,17 @@ The following options can optionally be configured:
 
+**User defined environment variables**
+
+`env` makes it possible to pass arbitrary environment variables to the buildpacks.
```
_pass_ makes it sounds like we expose variables in the user's current environment. This is more constrained, right?
```suggestion
`env` makes it possible to configure specific environment variables for buildpacks.
Many buildpacks use environment variables to adjust their detection and the build phases,
such as selecting specific versions of language runtimes.
Note that user's current environment is not passed through to buildpacks.
```

dgageot replied:
I'll fo that

briandealwis commented on docs/content/en/docs/pipeline-stages/builders/buildpacks.md:
```diff
@@ -25,9 +25,17 @@ The following options can optionally be configured:
+
 **Example**
 
 The following `build` section, instructs Skaffold to build a
```
Could you also add an example of `env` — maybe with a go-template? — to the `samples/builders/buildpacks.yaml`?  (Is it possible to expose a value from the user's environment with this mechanism? `USER={{.USER}}`?)

(I likely don't know what I'm talking about here: I know only enough go templating to process `docker images —format`.)

dgageot replied:
That's a good idea to add such value to the sample. But I don't see any value that would really be useful. Any idea?

dgageot replied:
Sorry, with regards to nested templates, I still don't understand what your point is.

briandealwis replied:
`NPM_CONFIG_LOGLEVEL`?

dgageot replied:
Add GOPROXY

nkubala commented on pkg/skaffold/build/cache/hash.go:
```diff
@@ -115,6 +125,16 @@ func retrieveBuildArgs(artifact *latest.Artifact) map[string]*string {
+	default:
+		return nil
+	}
+}
```
why use a switch statement here?
```golang
if artifact.BuildpackArtifact != nil {
  return artifact.BuildpackArtifact.Env
}
return nil
```

dgageot replied:
I'll fix that

dgageot commented:
@balopat @nkubala @briandealwis It should be ready for a new review.

briandealwis commented on integration/examples/buildpacks/skaffold.yaml:
```diff
@@ -7,3 +7,4 @@ build:
       builder: "heroku/buildpacks"
       env:
       - GOVERSION=1.13.1
+      - GOPROXY={{.GOPROXY}}
```
Cool!  I didn't realize it was that easy!
